tests/jdom/adapters.spec.ts

```diff
@@ -0,0 +1,153 @@
+import { suite, test, afterEach } from "mocha"
+import { JDBus } from "../../src/jdom/bus"
+import JDServiceProvider from "../../src/jdom/serviceprovider"
+import { EVENT, SRV_BUTTON, SRV_BUTTON_GESTURE, DEVICE_ANNOUNCE, ROLE_BOUND } from "../../src/jdom/constants";
+import { mkBus } from "../testutils";
+
+import ButtonGestureAdapter, { AdapterServer } from "../../src/servers/buttongestureadapter"
+import ButtonServer from "../../src/servers/buttonserver"
+import { JDEvent } from "../../src/jdom/event";
+import { JDDevice } from "../../src/jdom/device";
+import RoleManager from "../../src/servers/rolemanager"
+import JDServiceServer from "../../src/jdom/serviceserver";
+import { assert } from "../../src/jdom/utils";
+import { JDService } from "../../src/jdom/service";
+import { RoleManagerClient } from "../../src/jdom/rolemanagerclient";
+import { instanceOf } from "prop-types";
+
+interface BusDevice {
+    server: JDServiceServer,
+    roleName?: string,
+}
+
+// how the heck is this not a native operation
+function setEquals<T>(set1: Set<T>, set2: Set<T>): boolean {
+    if (set1.size != set2.size) {
+        return false
+    }
+    set1.forEach(value => {
+        if (!set2.has(value)) {
+            return false
+        }
+    })
+    return true
+}
+
+// Creates a bus with the specified servers, bound to the specified roles.
+// Returns once all devices are registered, and adapters are ready.
+// TODO should a timeout live here? or should that be a higher level responsibility?
+async function createBus(busDevices: BusDevice[]): Promise<JDBus> {
+    const bus = mkBus()
+
+    const serverDeviceRoleList: [JDServiceServer, JDDevice, string?][] = busDevices.map(busDevice => {
+        const device = bus.addServiceProvider(new JDServiceProvider(
+            [busDevice.server]
+            ))  // TODO support multi-service devices?
+        return [busDevice.server, device, busDevice.roleName]
+    })
+
+    // Wait for created devices to be announced, so services become available
+    // TODO is this a good way to write async code in TS?
+    await new Promise((resolve, reject) => {
```

Comment:
This is handled by the role manager.

It emits BOUND event when all roles are bound (and UNBOUND when some are lost)

Reply:
Is it? Don't I need the JDService (rather than the JDServiceServer) to set the roles in the first place?